### ISSUE TITLE
Question: dataloader on N to N relationship?

### ISSUE THREAD

**Issue by jjangga0214** [graphql/dataloader#201]
Hi,

Using dataloader for solving N+1 problem when dealing with N to 1 relationship is straightforward.  

However, what about N to N relationship?   

Let's assume there are `Group` and `User` tables in RDB. A user can be a member of multiple groups. A group has multiple users.   

```js
const resolver = {
  Group: {
    members: ({ id }) => groupUserLoader.load(id)
  }
}
```

```js  
// This is pseudo code. 
// Handling retrieved rows and returning [[User]] is omitted for brevity.
// `user_group` table is a bridge table between `user` and `group`.
const groupUserLoader = new DataLoader( groupIds =>
  sql(`SELECT user.* 
       FROM user, user_group 
       WHERE user_group.groupId IN ${groupIds}
       AND user.id = user_group.userId`)
}
```

Is this a proper approach? Or is there a better solution?

I see this makes N to N batch cumbersome. I need to declare `groupUserLoader`, `fooUserLoader`, `barUserLoader` ... not just `userLoader` once.

**P.S.** Should I refer this 1 to N relationship in the viewpoint of resolvers, not the schema?

**Comment by tuananh** [comment]
I'm interested to hear what others are doing reg. this as well

**Comment by jaredjj3** [comment]
From the docs:

>A batch loading function accepts an Array of keys, and returns a Promise which resolves to an Array of values. There are a few constraints that must be upheld:
> - The Array of values must be the same length as the Array of keys.
> - Each index in the Array of values must correspond to the same index in the Array of keys.

Your batch loading function will not uphold those constraints.

A way of upholding the constraints is to create a separate function that knows how to map a many-to-many relationship.

```typescript
type Key = string | number;

type KeysGetter<V> = (value: V) => Key[];

type UniqueIdentifierGetter<V, T> = (value: V) => T;

interface ValuesByKey<V> {
  [key: Key]: V;
}

function alignManyToMany<V>(
  keys: Key[],
  values:  V[],
  getKeysFromValue: KeysGetter<V>,
  getUniqueIdentifier: UniqueIndentifierGetter<V, number>
) {
  // From the values you got from your database, compute what key they correspond to.
  // Note that each key can have multiple values and each value can have multiple keys.
  const valuesByKey = values.reduce<ValuesByKey<V>>((obj, value) => {
    const computedKeys = getKeysFromValue(value)
    for (const computedKey of computedKeys) {
      obj[computedKey] = obj[computedKey] || [];
      obj[computedKey].push(value);
    }
    return obj;
  }, {});

  // Align the values with the keys by iterating over the keys and populating
  // the aligned array. Keys that are not in the valuesByKey object will end up
  // having an empty array. You can choose to handle missing values the way
  // your use case requires.  After this for-loop, we've satisfied both constraints
  // that the Dataloader library requires.
  const alignedValues = new Array(keys.length);
  for (const [ndx, key] of keys.entries()) {
    alignedValues[ndx] = valuesByKey[key] || [];
  }

  // As a bonus, ensure that each array in the alignedValues array contains unique
  // elements. Due to the many-to-many relationship, each key will have many values 
  // associated with it. Therefore, we cannot use the key to dedup values. I could 
  // have alternatively deduped when constructing the valuesByKey object.
  const dedupedAlignedValues = new Array(keys.length);
  for (const [ndx, values] of alignedValues.entries()) {
    dedupedAlignedValues[ndx] = [];
    const seen = new Set<V>();
    for (const value of values) {
      const id = getUniqueIdentifier(value);
      if (seen.has(id)) continue;
      seen.add(id);
      dedupedAlignedValues[ndx].push(value);
    }
  }

  return depuedAlignedValues;
}
```
I implemented this imperatively to illustrate what should go on behind-the-scenes. I encourage you to consider this as an incomplete solution until you evaluate if it fits your use case(s) and refactor it (probably debug it, too).

What are your thoughts?

**Comment by jjangga0214** [comment]
@jaredjj3 Thank you for sharing your idea :)

However, it seems like you misunderstood my point (this might be due to my unclear description.).
Let me explain the it more clearly.    

Assume there are `Group`, `User`, `Foo`, `Bar`. All of them are N to N each other. 
if I only need simple "primary key loaders", which takes PK of return type, then 4 dataLoaders would be enough (`userLoader`, `groupLoader`, `fooLoader`, `barLoader`). But actually I should declare many more if "foreign key loaders" (e.g. look at `groupUserLoader`) are needed. Total 12 "foreign key loaders" more! (e.g. `groupUserLoader`, `fooUserLoader`, `barUserLoader` for User. `userGroupLoader`, `fooGroupLoader`, `barGroupLoader` for Group and so on.)

As schema grows and types gets more connected, "foreign key loaders" increase with a phase of n^2 where n is the number of types on schema. Look at I already have 16 (= 4 pk loaders + 12 fk loaders) dataLoaders only for `Group`, `User`, `Foo`, `Bar`.

**Comment by jaredjj3** [comment]
@jjangga0214, I understand the problem now. My apologies.

I think that your scaling problem is inherent to the nature of your schema. If you must associate `n` tables with `m` tables in many-to-many relationships, then you must have a mechanism to load the join tables between them.

I imagine that your fk loaders look like the query in your original post. What if you created a function that took let's say an `AssociationCtx` and returned a batch loading function?

Just brainstorming:

```typescript
interface AssociationCtx {
  primaryTable: string;
  primaryKey: string;
  joinTable: string;
  foreignTable: string;
  foreignKey: string;
}

type BatchLoadingFn = (keys: any) => any[];

function getManyToManyBatchFn (ctx: AssociationCtx): BatchLoadingFn {
  return function(keys) {
    const records = sql(`SELECT ${ctx.primaryTable}.* 
       FROM ${ctx.primaryTable}, ${ctx.foreignTable} 
       WHERE ${ctx.joinTable}.${ctx.foreignKey} IN ?
       AND ${ctx.primaryTable}.id = ${ctx.joinTable}.${ctx.primaryKey}`, keys);

    // code to uphold batch function constraints
  }
}
```

While this still requires you to declare your many-to-many relationships, it alleviates the boilerplate code associated with it.

You could take it a step further and create a DSL that would loop through your associations and call `getManyToManyBatchFn` for each many-to-many association. However, I feel that this level of abstraction could be a breeding ground for subtle bugs - I'm not sure.

This approach assumes that all your fk loaders look the same and _will_ look the same. What's your feedback?

**Comment by jjangga0214** [comment]
@jaredjj3 Good! 
If a schema has strictly conventional(e.g. fk naming, junction table naming) relationships among tables, it will work.

However, updating cache for other dataLoaders wouldn't be applicable by this way, unless a technique like a language reflection is applied. For example, the code below should be included in `groupUserLoader`.

```js
  for (let users of rows) {
    for (let user of user) {
      userLoader.prime(user.id, user)
    }
  } 
```

**Comment by leebyron** [comment]
Closing this issue since it seems the original question was answered.

**Comment by jjangga0214** [comment]
>  it seems the original question was answered.

@leebyron Well, not really. 

My point was that the number of dataloaders increase exponentially with N to N (or 1 to N (?)) relationship. @jaredjj3 suggested a way to conveniently initialize many dataloaders with only by one function, but it does not solve cache problem.

**Comment by MattGson** [comment]
To solve the cache issue you need to separately load the joins (edges) and the entities (nodes) themselves.

So to load all the books for a library you would have:

libraryByIdLoader, bookByIdLoader, and bookIdByLibraryLoader.
Then to load all the books for the library you call:

library = LibraryByIdLoader.load(id);
bookIds = bookIdsByLibraryLoader.load(library.id);
books = bookByIdLoader.loadMany(bookIds);

The advantage of this is that you have a single source of truth for how to load each model.
That source of truth will always have the full cache for that object type (do note that the caching benefit is by far smaller than the batching in most cases I've seen).
Each many-to-many association can only be loaded in at most 2 ways so then you have at most < 2N loaders where n is the total number of tables in your db. Using a system like the one suggested above, you can create most of these loaders with a single line of code. At most two lines of code per table (only one for entity tables, two for edges). I don't see the number of loaders as a problem here. If you want a small number of optimised nested queries then you shouldn't need data loader anyway.
